distribution/lib/Standard/Base/0.0.0-dev/src/Data/Json.enso

```diff
@@ -281,3 +104,148 @@ type Marshalling_Error
             "Type mismatch error: the json with type `" + json_text + "` did not match the format `" + format_text + "`."
         Marshalling_Error.Missing_Field _ field _ ->
             "Missing field in Json: the field `" + field.to_text "` was missing in the json."
+
+type JS_Object
+    ## Creates a JS_Object from a list of key-value pairs.
+       Keys must be `Text` values.
+       Values will be recursively converted to JSON serializable as needed.
+    from_pairs : Vector -> JS_Object
+    from_pairs pairs =
+        js_object = pairs.fold new_object current->pair->
+            case pair.at 0 of
+                text : Text ->
+                    set_value current text (pair.at 1).to_js_object
+                    current
+                _ -> Error.throw (Illegal_Argument.Error "JS_Object.from_pairs: key must be a Text value")
+        JS_Object.Value js_object
+
+    ## PRIVATE
+    Value js_object
+
+    ## Get a value for a key of the object.
+
+       Arguments:
+       - key: The key to get.
+       - other: The value to return if the key is not found.
+    get : Text -> Any -> JS_Object | Boolean | Number | Nothing | Text | Vector
+    get self key ~other=Nothing =
+        if (has_property self.js_object key) . not then other else
+            value = get_value self.js_object key
+            make_enso value
+
+    ## Get a value for a key of the object.
+       If the key is not found, throws a `No_Such_Key` error.
+
+       Arguments:
+       - key: The key to get.
+    at : Text -> JS_Object | Boolean | Number | Nothing | Text | Vector
+    at self key = self.get key (Error.throw (No_Such_Key.Error self key))
+
+    ## Get the keys of the object
+    field_names : Vector
+    field_names self =
+        Vector.from_polyglot_array (get_property_names self.js_object)
+
+    ## Gets the number of keys in the object
+    length : Number
+    length self =
+        get_property_names self.js_object . length
+
+    ## Convert the object to a Vector of Pairs
+    to_vector : Vector
+    to_vector self =
+        keys = get_property_names self.js_object
+        proxy = Array_Proxy.new keys.length (i-> Pair.new (keys.at i) (self.get (keys.at i)))
+        Vector.from_polyglot_array proxy
+
+    ## Convert the object to a JSON
+    to_js_object : JS_Object
+    to_js_object self = self
+
+    ## Convert to string
+    to_text : Text
+    to_text self = Json.stringify self
+
+    ## Checks if this JS_Object is equal to another JS_Object.
+
+       Arguments:
+       - that: The map to compare `self` to.
+    == : JS_Object -> Boolean
```

Comment:
I hypothesize that this is incorrect.

Let's have two objects:
```
obj1 = {"foo": null}
obj2 = {"bar": 42}
? obj1 == obj2
```

IMO it will yield true.
Key lengths are the same.
We will iterate thru self keys. `obj2.get "foo"` will give Nothing for the missing field. But that will be equal to the Nothing returned for the existing field bound to JS .

Moreover, with this bug equality is not symmetric, because this time correctly `obj2 == obj1` will be false.

Reply:
good point. Will fix that